Mathlib/CategoryTheory/Limits/Comma.lean

```diff
@@ -0,0 +1,307 @@
+      naturality := fun j₁ j₂ t => by
+        ext
+        . simp [← c₁.w t]
+        . simp [← c₂.w t] }
+#align category_theory.comma.cocone_of_preserves CategoryTheory.Comma.coconeOfPreserves
+
+/-- Provided that `L` preserves the appropriate colimit, then the cocone in `coconeOfPreserves` is
+a colimit. -/
+def coconeOfPreservesIsColimit [PreservesColimit (F ⋙ fst L R) L] {c₁ : Cocone (F ⋙ fst L R)}
+    (t₁ : IsColimit c₁) {c₂ : Cocone (F ⋙ snd L R)} (t₂ : IsColimit c₂) :
+    IsColimit (coconeOfPreserves F t₁ c₂)  where
+  desc s :=
+    { left := t₁.desc ((fst L R).mapCocone s)
+      right := t₂.desc ((snd L R).mapCocone s)
+      w :=
+        (isColimitOfPreserves L t₁).hom_ext fun j =>
+          by
+          rw [coconeOfPreserves_pt_hom, (isColimitOfPreserves L t₁).fac_assoc,
+            colimitAuxiliaryCocone_ι_app, assoc, ← R.map_comp, t₂.fac, L.mapCocone_ι_app, ←
+            L.map_comp_assoc, t₁.fac]
+          exact (s.ι.app j).w }
+  uniq s m w := by
+    apply CommaMorphism.ext
+    . exact t₁.uniq ((fst L R).mapCocone s) _ (fun j => by simp [← w])
+    . exact t₂.uniq ((snd L R).mapCocone s) _ (fun j => by simp [← w])
+#align category_theory.comma.cocone_of_preserves_is_colimit CategoryTheory.Comma.coconeOfPreservesIsColimit
+
+instance hasLimit (F : J ⥤ Comma L R) [HasLimit (F ⋙ fst L R)] [HasLimit (F ⋙ snd L R)]
+    [PreservesLimit (F ⋙ snd L R) R] : HasLimit F :=
+  HasLimit.mk ⟨_, coneOfPreservesIsLimit _ (limit.isLimit _) (limit.isLimit _)⟩
+#align category_theory.comma.has_limit CategoryTheory.Comma.hasLimit
+
+instance hasLimitsOfShape [HasLimitsOfShape J A] [HasLimitsOfShape J B]
+    [PreservesLimitsOfShape J R] : HasLimitsOfShape J (Comma L R) where
+#align category_theory.comma.has_limits_of_shape CategoryTheory.Comma.hasLimitsOfShape
+
+instance hasLimits [HasLimits A] [HasLimits B] [PreservesLimits R] : HasLimits (Comma L R) :=
+  ⟨fun _ _ => inferInstance⟩
+#align category_theory.comma.has_limits CategoryTheory.Comma.hasLimits
+
+instance hasColimit (F : J ⥤ Comma L R) [HasColimit (F ⋙ fst L R)] [HasColimit (F ⋙ snd L R)]
+    [PreservesColimit (F ⋙ fst L R) L] : HasColimit F :=
+  HasColimit.mk ⟨_, coconeOfPreservesIsColimit _ (colimit.isColimit _) (colimit.isColimit _)⟩
+#align category_theory.comma.has_colimit CategoryTheory.Comma.hasColimit
+
+instance hasColimitsOfShape [HasColimitsOfShape J A] [HasColimitsOfShape J B]
+    [PreservesColimitsOfShape J L] : HasColimitsOfShape J (Comma L R) where
+#align category_theory.comma.has_colimits_of_shape CategoryTheory.Comma.hasColimitsOfShape
+
+instance hasColimits [HasColimits A] [HasColimits B] [PreservesColimits L] :
+    HasColimits (Comma L R) :=
+  ⟨fun _ _ => inferInstance⟩
+#align category_theory.comma.has_colimits CategoryTheory.Comma.hasColimits
+
+end Comma
+
+namespace Arrow
+
+instance hasLimit (F : J ⥤ Arrow T) [i₁ : HasLimit (F ⋙ leftFunc)] [i₂ : HasLimit (F ⋙ rightFunc)] :
+    HasLimit F := by
+  haveI : HasLimit (F ⋙ Comma.fst _ _) := i₁
```

Comment:
I made a [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Type.20class.20inference.20in.20category.20theory) thread about this